File: ruby/cipher/hex_encoder.rb
module Cipher
  class HexEncoder
    attr_reader :inner

    def initialize(cipher)
      @inner = cipher
    end

    def encrypt(plaintext)
      bytes_to_hex(@inner.encrypt(plaintext.unpack('C*')))
    end

    def decrypt(ciphertext)
      @inner.decrypt(hex_to_bytes(ciphertext)).pack('C*')
    end

    private

    def bytes_to_hex(bytes)
      bytes.map{ |byte| "%02x" % byte }.join
    end

    def hex_to_bytes(hex)
      hex.scan(/.{2}/).map{ |char| char.hex }
    end
  end
end

require_relative './null_cipher.rb'
require 'test/unit'

class TestHexEncoder < Test::Unit::TestCase
  def test_encryption
    cipher = Cipher::HexEncoder.new(Cipher::NullCipher.new)

    assert_equal(cipher.encrypt('Burning'), '4275726e696e67')
  end

  def test_decryption
    cipher = Cipher::HexEncoder.new(Cipher::NullCipher.new)

    assert_equal(cipher.decrypt('4275726e696e67'), 'Burning')
  end
end

File: src/cipher/hex_encoder.rs
extern crate rustc_serialize;

use self::rustc_serialize::hex::{ FromHex, ToHex };

use cipher::{ Cipher, Encoder };

pub struct HexEncoder<C: Cipher> {
    cipher: C,
}

impl<C> HexEncoder<C> where C: Cipher {
    pub fn new(cipher: C) -> HexEncoder<C> {
        HexEncoder {
            cipher: cipher
        }
    }
}

impl<C: Cipher> Encoder for HexEncoder<C> {
    fn encrypt(&self, plaintext: &str) -> String {
        self.cipher.encrypt(plaintext.bytes().collect()).to_hex()
    }

    fn decrypt(&self, ciphertext: &str) -> String {
        String::from_utf8(self.cipher.decrypt(ciphertext.from_hex().unwrap())).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use cipher::{ Encoder, HexEncoder, NullCipher };

    #[test]
    fn it_can_encode_hex_correctly() {
        let encoder = HexEncoder::new(NullCipher::new());

        assert_eq!(encoder.encrypt("Burning"), "4275726e696e67");
    }

    #[test]
    fn it_can_decode_hex_correctly() {
        let encoder = HexEncoder::new(NullCipher::new());

        assert_eq!(encoder.decrypt("4275726e696e67"), "Burning");
    }
}
